Upstream-Status: Inappropriate [configuration]

--- a/crypto/Makefile
+++ b/crypto/Makefile
@@ -108,7 +108,7 @@ $(LIB):	$(LIBOBJ)
 
 shared: buildinf.h lib subdirs
 	if [ -n "$(SHARED_LIBS)" ]; then \
-		(cd ..; $(MAKE) $(SHARED_LIB)); \
+		(cd ..; $(MAKE) -e $(SHARED_LIB)); \
 	fi
 
 libs:
--- a/Makefile.org
+++ b/Makefile.org
@@ -310,7 +310,7 @@ libcrypto$(SHLIB_EXT): libcrypto.a fips_
 
 libssl$(SHLIB_EXT): libcrypto$(SHLIB_EXT) libssl.a
 	@if [ "$(SHLIB_TARGET)" != "" ]; then \
-		$(MAKE) SHLIBDIRS=ssl SHLIBDEPS='-lcrypto' build-shared; \
+		$(MAKE) -e SHLIBDIRS=ssl SHLIBDEPS='-lcrypto' build-shared; \
 	else \
 		echo "There's no support for shared libraries on this platform" >&2; \
 		exit 1; \
--- a/ssl/Makefile
+++ b/ssl/Makefile
@@ -62,7 +62,7 @@ lib:	$(LIBOBJ)
 
 shared: lib
 	if [ -n "$(SHARED_LIBS)" ]; then \
-		(cd ..; $(MAKE) $(SHARED_LIB)); \
+		(cd ..; $(MAKE) -e $(SHARED_LIB)); \
 	fi
 
 files:
